diff --git a/google3/third_party/libsrtp/test/srtp_driver.c b/google3/third_party/libsrtp/test/srtp_driver.c
index 9bd4c66..f6a84f9 100644
--- a/google3/third_party/libsrtp/test/srtp_driver.c
+++ b/google3/third_party/libsrtp/test/srtp_driver.c
@@ -138,7 +138,7 @@ debug_module_t mod_driver = {
 
 int
 main (int argc, char *argv[]) {
-  char q;
+  int q;
   unsigned do_timing_test    = 0;
   unsigned do_rejection_test = 0;
   unsigned do_codec_timing   = 0;
@@ -169,13 +169,13 @@ main (int argc, char *argv[]) {
     if (status) {
     printf("error: load of srtp_driver debug module failed "
            "with error code %d\n", status);
-    exit(1);   
+    exit(1);
   }
 
   /* process input arguments */
   while (1) {
     q = getopt_s(argc, argv, "trcvld:");
-    if (q == -1) 
+    if (q == -1)
       break;
     switch (q) {
     case 't':
